File: 2555-maximize-win-from-two-segments/2555-maximize-win-from-two-segments.cpp
class Solution {
public:
    int maximizeWin(vector<int>& pos, int k) {

        int n = pos.size(), w1 = 0, res = 0, prev = 0, next = 0;

        for (int i = 0; i < n; i++){
            while(next < n and pos[next]<=pos[i] + k){
                next++;
            }
            while(pos[prev]<pos[i] -k) prev++;
            res = max(res, w1+next-i);
            w1 = max(w1, i-prev+1);   
        }
    return res;
    }
};

File: 2555-maximize-win-from-two-segments/2555-maximize-win-from-two-segments.py
class Solution:
    def maximizeWin(self, pos: List[int], k: int) -> int:
        
        n = len(pos)
        res = 0
        seg1 = 0
        prev = 0
        next_ = 0

        for i in range(n):
            while (next_ < n) and (pos[next_] - pos[i] <= k):
                next_+=1
            while (prev < n) and (pos[i] - pos[prev]) > k:
                prev+=1
            res = max(res,seg1 + next_ - i )
            seg1 = max(seg1, i - prev + 1)
        return res
